### PR TITLE
[TACHYON-833] Implement Journal crash test out of unit test

### PR DESCRIPTION
**Pull request by swatDong**
This PR implements a util to test Journal when crashing. The test can be run by `bin/tachyon jCrashTest [args]`. Following is the workflow:

    i. Start Master and Clients to request to Master.
    ii. Kill Master.
    iii. Restart Master and check status.

This test is at client side and is put into the examples package. I'm going to add more operations and configurable settings.

### CODE REVIEW

**Comment by gpang** [comment]
Jenkins, test this please.

**Comment by apc999** [comment]
Jenkins, test this please.

File: bin/tachyon [hunk 1]
```diff
@@ -16,6 +16,7 @@ function printUsage {
   echo -e "  loadufs       \t Load existing files in underlayer filesystem into Tachyon."
   echo -e "  runTest       \t Run a end-to-end test on a Tachyon cluster."
   echo -e "  runTests      \t Run all end-to-end tests on a Tachyon cluster."
+  echo -e "  jCrashTest    \t Test the Master Journal System in a crash scenario"
```
Comment (gpang) [inline bin/tachyon]: This should actually say "journal" instead of just "j". Also, does it make sense to have this new command, or would it fit under runTest?

Nevermind, it looks like it stops existing server and creates a new one. That info should be in the "help" of the journal test.

File: examples/pom.xml [hunk 2]
```diff
@@ -19,6 +19,11 @@
 
   <dependencies>
     <dependency>
+      <groupId>commons-cli</groupId>
```
Comment (gpang) [inline examples/pom.xml]: Does tachyon already depend on commons-cli elsewhere?

File: examples/src/main/java/tachyon/examples/JournalCrashTest.java [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+
+import tachyon.Constants;
+import tachyon.TachyonURI;
+import tachyon.Version;
+import tachyon.client.TachyonFS;
+import tachyon.conf.TachyonConf;
+import tachyon.util.CommonUtils;
+
+/**
+ * Class to perform Journal crash test. Launch clients to generate the Journal and check status.
```
Comment (gpang) [inline examples/src/main/java/tachyon/examples/JournalCrashTest.java]: the clients issue commands to the master, and the master generates journal events.

File: examples/src/main/java/tachyon/examples/JournalCrashTest.java [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+
+  /**
+   * The client thread class. Each thread hold a Tachyon Client and keep requesting to Master.
+   */
+  static class ClientThread implements Runnable {
+    /** Which type of operation this thread should do. */
+    private final ClientOpType mOpType;
+    /** The Tachyon Client hold by this thread. */
+    // TODO: use TachyonFileSystem instead of the deprecated TachyonFS
+    private final TachyonFS mTfs;
```
Comment (gpang) [inline examples/src/main/java/tachyon/examples/JournalCrashTest.java]: Can you use the new client?

File: examples/src/main/java/tachyon/examples/JournalCrashTest.java [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+            if (!mTfs.rename(fid, new TachyonURI(mWorkDir + mSuccessNum + "-rename"))) {
+              break;
+            }
+          }
+          //else if (ClientOpType.CREATE_TABLE == mOpType) {
+          //  if (mTfs.createRawTable(new TachyonURI(mWorkDir + mSuccessNum), 1) == -1) {
+          //    break;
+          //  }
+          //}
+          mSuccessNum++;
```
Comment (gpang) [inline examples/src/main/java/tachyon/examples/JournalCrashTest.java]: please add a space before the "++".

File: examples/src/main/java/tachyon/examples/JournalCrashTest.java [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+          CommonUtils.sleepMs(null, 100);
+        }
+      }  catch (Exception e) {
+        // Something crashed. Stop the thread.
+      }
+    }
+  }
+
+
```
Comment (gpang) [inline examples/src/main/java/tachyon/examples/JournalCrashTest.java]: extra blank line?

File: examples/src/main/java/tachyon/examples/JournalCrashTest.java [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+      obOut.writeObject(mOpTypes);
+      obOut.writeObject(mSuccessNums);
+      obOut.writeObject(mWorkDirs);
+      obOut.close();
+    }
+  }
+
+  // The two Exit Codes are used to tell script if the test runs well.
+  private static final int EXIT_FAILED = 99;
+  private static final int EXIT_SUCCESS = 100;
```
Comment (gpang) [inline examples/src/main/java/tachyon/examples/JournalCrashTest.java]: why are these status codes this way? Why not -1 and 0 or something more standard?

File: examples/src/main/java/tachyon/examples/JournalCrashTest.java [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private static int sClientNum = 10;
+  private static TachyonURI sMasterAddress = null;
+  private static String sTestDir = null;
+  private static List<Thread> sClientThreadList = null;
+  private static List<ClientThread> sClientThreads = null;
```
Comment (gpang) [inline examples/src/main/java/tachyon/examples/JournalCrashTest.java]: what is the difference between sClientThreads and sClientThreadList?

File: examples/src/main/java/tachyon/examples/JournalCrashTest.java [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+    }
+    tfs.close();
+    return true;
+  }
+
+  public static void main(String[] args) {
+    if (args.length < 4) {
+      System.out.println("java -cp tachyon-" + Version.VERSION + "-jar-with-dependencies.jar "
+          + "tachyon.examples.JournalCrashTest " + "<0:Run Clients|1: Check Status> "
+          + "<TachyonMasterAddress> <TestTachyonDir> <TmpOutFile> [-options]");
```
Comment (gpang) [inline examples/src/main/java/tachyon/examples/JournalCrashTest.java]: why are some options in the command line, and some in "-options"?

File: examples/src/main/java/tachyon/examples/JournalCrashTest.java [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+      System.exit(EXIT_FAILED);
+    }
+
+    int runType = Integer.parseInt(args[0]);
```
Comment (gpang) [inline examples/src/main/java/tachyon/examples/JournalCrashTest.java]: Why are there two different modes? Why doesn't the test run the clients, and then check the status at the end?

**Comment by gpang** [comment]
Thanks for this test! I left some comments in the code.

File: bin/tachyon (hunk 1 above)
Comment (swatDong) [inline bin/tachyon]: I think it should be separated with runTest because this test needs to start/stop Master.
I'll change it to 'journalCrashTest' and update the 'help'.

File: examples/pom.xml (hunk 2 above)
Comment (swatDong) [inline examples/pom.xml]: No. I haven't found commons-cli in Tachyon's pom or jar files.

File: examples/src/main/java/tachyon/examples/JournalCrashTest.java (hunk 3 above)
Comment (swatDong) [inline examples/src/main/java/tachyon/examples/JournalCrashTest.java]: Thanks. I'll update the comments.

File: examples/src/main/java/tachyon/examples/JournalCrashTest.java (hunk 4 above)
Comment (swatDong) [inline examples/src/main/java/tachyon/examples/JournalCrashTest.java]: Yes. I'll fix the current issues, address the comments, then use the new client.
After that I'll remove the 'WIP' tag of this PR.

File: examples/src/main/java/tachyon/examples/JournalCrashTest.java (hunk 7 above)
Comment (swatDong) [inline examples/src/main/java/tachyon/examples/JournalCrashTest.java]: The 'bin/tachyon' script uses `ret > 128` to determine if the process is terminated due to signal.
If the status code is -1 (`System.exit(-1)`), `ret` in script will be 255. The status codes here should be not greater than 128.

To be more standard, I'll use 0 for success and 1 for failed.

File: examples/src/main/java/tachyon/examples/JournalCrashTest.java (hunk 8 above)
Comment (swatDong) [inline examples/src/main/java/tachyon/examples/JournalCrashTest.java]: One is used to wait all the threads finished (`thread.join()`). Another is used to get the result (`clientThread.getSuccessNum()`).

I can use `ClientThread extends Thread` to only have one list.

File: examples/src/main/java/tachyon/examples/JournalCrashTest.java (hunk 9 above)
Comment (swatDong) [inline examples/src/main/java/tachyon/examples/JournalCrashTest.java]: My previous idea is to put the optional args in "-options".
I'll put all the args in "-options" with some default values.

File: examples/src/main/java/tachyon/examples/JournalCrashTest.java (hunk 10 above)
Comment (swatDong) [inline examples/src/main/java/tachyon/examples/JournalCrashTest.java]: Between the two modes there is a step to kill the Master, which should be done by script.
If needed, I can make the test be aware of Master crash and run the test in one mode.

**Comment by swatDong** [comment]
Two more steps:
i. make all the args in "-options"
ii. use the new tachyon.client.file.TachyonFileSystem

**Comment by swatDong** [comment]
@gpang I've addressed the comments. And make this test
  1) run clients and check status in one program
  2) put all options in [-options]

File: bin/tachyon [hunk 11]
```diff
@@ -116,6 +117,59 @@ function runTest {
   return $RET
 }
 
+function journalCrashTest {
+  if [ "$#" -lt 1 ]; then
+    echo "Usage: tachyon journalCrashTest <test-duration-in-seconds> [-options]"
```
Comment (gpang) [inline bin/tachyon]: Is there a way to leverage the commons-cli library to print out the usage automatically? Having it print out in both this file and the .java file seems error prone to keep consistent.

Also, why is the test duration not in the cli options?

File: bin/tachyon [hunk 12]
```diff
@@ -116,6 +117,59 @@ function runTest {
...
+  done
+  echo "Tachyon Master Killed. Check status..."
+  wait $testPid
+  RET=$?
+
+  echo "Stop Tachyon Master"
+  $bin/tachyon-stop.sh
+  # If test process received a signal abort early
+  # Exit Codes for signals terminating a JVM are 128 + signal
+  if [ $RET -gt 128 ]; then
```
Comment (gpang) [inline bin/tachyon]: Does this line still work when the status changed to 0 and 1?

File: examples/src/main/java/tachyon/examples/JournalCrashTest.java [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+import tachyon.client.file.TachyonFile;
+import tachyon.client.file.TachyonFileSystem;
+import tachyon.conf.TachyonConf;
+import tachyon.util.CommonUtils;
+
+/**
+ * Class to perform Journal crash test. The clients issue commands to the master, and the master
+ * generates journal events. Check if the master can generate and reproduce the journal correctly.
+ */
+public class JournalCrashTest {
```
Comment (gpang) [inline examples/src/main/java/tachyon/examples/JournalCrashTest.java]: Probably not in this PR, but it would be great to have this test kill/restart the master multiple times during the duration of the test.

File: examples/src/main/java/tachyon/examples/JournalCrashTest.java [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+      try {
+        thread.join();
+      } catch (InterruptedException e) {
+        LOG.error("Error when waiting thread", e);
+      }
+    }
+
+    // Restart Tachyon Master.
+    String restartMasterCommand = new TachyonConf().get(Constants.TACHYON_HOME)
+        + "/bin/tachyon-start.sh master";
```
Comment (gpang) [inline examples/src/main/java/tachyon/examples/JournalCrashTest.java]: It is a little confusing that the master is started in the bash script, but also started here. Why is the master started in both places?

Ideally, it should only be started/killed in one place.

File: examples/src/main/java/tachyon/examples/JournalCrashTest.java [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+    }
+    sTestDir = cmd.getOptionValue("testDir", "/default_tests_files");
+    sCreateFileClientNum = Integer.parseInt(cmd.getOptionValue("creates", "2"));
+    sCreateDeleteClientNum = Integer.parseInt(cmd.getOptionValue("deletes", "2"));
+    sCreateRenameClientNum = Integer.parseInt(cmd.getOptionValue("renames", "2"));
+    return true;
+  }
+
+  private static void printUsage() {
+    System.out.println("Usage: java -cp tachyon-" + Version.VERSION + "-jar-with-dependencies.jar "
```
Comment (gpang) [inline examples/src/main/java/tachyon/examples/JournalCrashTest.java]: Is there an automatic way for the options to print out the usage, instead of having to do it manually?

File: bin/tachyon (hunk 11 above)
Comment (swatDong) [inline bin/tachyon]: I'll use HelpFormatter.printHelp to print out the usage and only put it in the java file.
Then, I'll kill/restart Master for multiple times in the java file instead of a specified time duration.

File: bin/tachyon (hunk 12 above)
Comment (swatDong) [inline bin/tachyon]: Yes. If the test process is aborted by a signal, this line will work.

**Comment by swatDong** [comment]
Jenkins, test it please.

**Comment by swatDong** [comment]
Jenkins, test this please.

**Comment by swatDong** [comment]
@gpang I've addressed the comments and updated with master branch.
    1) Print the help in java by using `org.apache.commons.cli.HelpFormatter`
    2) Kill/restart Master only in java for multiple times

File: examples/src/main/java/tachyon/examples/JournalCrashTest.java [hunk 16]
```diff
@@ -0,0 +1,387 @@
...
+   * Parse the input args with a command line format, using
+   * <code>org.apache.commons.cli.CommandLineParser</code>.
+   * @param args the input args
+   * @return true if parse successfully and can run the next step, false if parse failed or need to
+   * print out the help information.
+   */
+  private static boolean parseInputArgs(String[] args) {
+    Options options = new Options();
+    options.addOption("help", false, "Show help for this test");
+    options.addOption("crashes", true, "Repeat times to kill/restart Master");
```
Comment (gpang) [inline examples/src/main/java/tachyon/examples/JournalCrashTest.java]: I think the way crashes are done now will not stress journal in a variety of situations, since the crashes all happen right after another. I think another way to do this is to define the total time to run the test for, and the maximum time a master should ever be alive. This means I could run the test for 24 hours, and if I want the master to be up for at most 1 hour, the test will kill the master after a random time (less than an hour). This means in the 24 hour test, I know the master would have been killed at least 24 times, but with random intervals.

**Comment by gpang** [comment]
I just had a suggestion for how the crashes work. Thanks!

File: examples/src/main/java/tachyon/examples/JournalCrashTest.java (hunk 16 above)
Comment (swatDong) [inline examples/src/main/java/tachyon/examples/JournalCrashTest.java]: That sounds good. I'll update to the random crash.

**Comment by swatDong** [comment]
Have made a better crash way. Now the master will be killed with random intervals, and check status for each crash.

**Comment by gpang** [comment]
LGTM
